pkg/internal/gcp/gcp.go

```diff
@@ -0,0 +1,164 @@
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gcp
+
+import (
+	"context"
+	"encoding/base64"
+	"fmt"
+	"strings"
+
+	"google.golang.org/api/secretmanager/v1"
+
+	"github.com/go-logr/logr"
+
+	smmeta "github.com/itscontained/secret-manager/pkg/apis/meta/v1"
+	smv1alpha1 "github.com/itscontained/secret-manager/pkg/apis/secretmanager/v1alpha1"
+	"github.com/itscontained/secret-manager/pkg/internal/store"
+
+	"google.golang.org/api/option"
+
+	corev1 "k8s.io/api/core/v1"
+
+	"k8s.io/apimachinery/pkg/types"
+
+	ctrlclient "sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ store.Client = &GCP{}
+
+type GCP struct {
+	kubeClient ctrlclient.Client
+	store      smv1alpha1.GenericStore
+	log        logr.Logger
+	client     *secretmanager.Service
+}
+
+func New(ctx context.Context, kubeClient ctrlclient.Client, store smv1alpha1.GenericStore, log logr.Logger) (store.Client, error) {
+	g := &GCP{
+		kubeClient: kubeClient,
+		store:      store,
+		log:        log,
+	}
+	err := g.newClient(ctx)
+	if err != nil {
+		log.Error(err, "could not create new gcp client")
+		return nil, err
+	}
+	return g, nil
+}
+
+func (g *GCP) GetSecret(_ context.Context, ref smv1alpha1.RemoteReference) ([]byte, error) {
+	version := "latest"
+	if ref.Version != nil {
+		version = *ref.Version
+	}
+	data, err := g.readSecret(*ref.Name, version)
+	if err != nil {
+		return nil, err
+	}
+	return data[*ref.Name], nil
+}
+
+func (g *GCP) GetSecretMap(_ context.Context, ref smv1alpha1.RemoteReference) (map[string][]byte, error) {
+	version := "latest"
+	if ref.Version != nil {
+		version = *ref.Version
+	}
+	return g.readSecret(*ref.Name, version)
+}
+
+func (g *GCP) readSecret(id, version string) (map[string][]byte, error) {
+	projectID := g.store.GetSpec().GCP.ProjectID
+	name := id
+	if !strings.HasPrefix(id, "projects/") && projectID != nil {
+		name = fmt.Sprintf("projects/%s/secrets/%s/versions/%s", *projectID, id, version)
+	}
+	resp, err := g.client.Projects.Secrets.Versions.Access(name).Do()
+	if err != nil {
+		return nil, err
+	}
+	data, err := base64.URLEncoding.DecodeString(resp.Payload.Data)
+	if err != nil {
+		return nil, err
+	}
+	return map[string][]byte{id: data}, nil
+}
+
+func (g *GCP) newClient(ctx context.Context) error {
+	g.log.V(1).Info("creating new gcp api client")
+	var err error
+	var clientOption option.ClientOption
+	spec := g.store.GetSpec().GCP
+	if spec.AuthSecretRef == nil {
+		g.log.V(1).Info("no authentication defined. using environment variables")
+		g.client, err = secretmanager.NewService(ctx)
+		if err != nil {
+			return err
+		}
+		return nil
+	}
+	if spec.AuthSecretRef.JSON != nil && spec.AuthSecretRef.File != nil {
```

Comment:
This logic should be in a validating webhook: ref https://github.com/itscontained/secret-manager/issues/6.

Should either remove or put a TODO to remove when adding validating webhook.

Reply:
added